src/librustc_mir/interpret/operand.rs

```diff
@@ -444,13 +444,27 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             Field(field, _) => self.operand_field(base, field.index() as u64)?,
             Downcast(_, variant) => self.operand_downcast(base, variant)?,
             Deref => self.deref_operand(base)?.into(),
-            Subslice { .. } | ConstantIndex { .. } | Index(_) => if base.layout.is_zst() {
+            ConstantIndex { .. } | Index(_) if base.layout.is_zst() => {
                 OpTy {
                     op: Operand::Immediate(Scalar::zst().into()),
                     // the actual index doesn't matter, so we just pick a convenient one like 0
                     layout: base.layout.field(self, 0)?,
                 }
-            } else {
+            }
+            Subslice { from, to, from_end } if base.layout.is_zst() => {
```

Comment:
I am kind of surprised that we have a special path here for ZST anyway. I don't remember seeing that. Why would the main code path not work for that?

Reply:
Looks like we do this since https://github.com/rust-lang/rust/pull/54032... @oli-obk this looks just wrong to me. The usual place arithmetic should work for slices. Unfortunately the comments don't explain what was wrong with the old code, do you remember?

Reply:
Oh wow that PR was a subtle change of Miri engine semantics in the middle for a PR about something entirely different... in the future I think we should avoid mixing concerns like that.

Reply:
:innocent:

Reply:
I'm opening a branch right now to look at this in more detail

Reply:
> Why would the main code path not work for that? ZST should only very rarely need any special-casing.

The reason we need this is this assert
https://github.com/rust-lang/rust/blob/c64eecf4d0907095928fb36fd3a1dd5fb2d9ff06/src/librustc_mir/interpret/operand.rs#L454-L456
triggering for ZSTs. The comment above the assert is wrong, we can reach this for empty arrays and arrays of ZSTs.

I can either fix this here by replacing the `assert_mem_place` with a match that produces a `MemPlace::dangling()` for ZSTs, or I can change https://github.com/rust-lang/rust/blob/de0abf7599023b71dd72b44f0165e86c040ee7ea/src/librustc_mir/interpret/place.rs#L224-L229 to do this, which seems strictly more ICE avoiding and looking at the use sites would actually help choose the better code path in a few situations.

Reply:
I opened a PR with the solution that changes `try_as_mplace` because it allows nice cleanups: https://github.com/rust-lang/rust/pull/67501

Reply:
Oh, this is `operand.rs`, not `place.rs`. Yeah that changes things.^^

Let's discuss in the new PR; thanks for opening that!